rectools/metrics/dq.py

```diff
@@ -0,0 +1,315 @@
+#  Copyright 2024 MTS (Mobile Telesystems)
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+"""Recommendations data quality metrics."""
+
+import typing as tp
+
+import attr
+import numpy as np
+import pandas as pd
+
+from rectools import Columns
+from rectools.metrics.base import MetricAtK
+from rectools.utils import fast_isin_for_sorted_test_elements, select_by_type
+
+
+@attr.s
+class _RecoDQMetric(MetricAtK):
+    """
+    Recommendations data quality metric base class.
+
+    Warning: This class should not be used directly.
+    Use derived classes instead.
+
+    Parameters
+    ----------
+    k : int
+        Number of items at the top of recommendations list that will be used to calculate metric.
+    deep: bool, default `False`
+        Whether to calculated detailed value of the metric for each user. Otherwise just the share of
+        users with identified problems will be returned (this is the default behaviour).
+    """
+
+    deep: bool = attr.ib(default=False)
+
+    def calc(self, reco: pd.DataFrame) -> float:
+        """
+        Calculate metric value.
+
+        Parameters
+        ----------
+        reco : pd.DataFrame
+            Recommendations table with columns `Columns.User`, `Columns.Item`, `Columns.Rank`.
+
+        Returns
+        -------
+        float
+            Value of metric (average between users).
+        """
+        per_user = self.calc_per_user(reco)
+        return per_user.mean()
+
+    def calc_per_user(self, reco: pd.DataFrame) -> pd.Series:
+        """
+        Calculate metric values for all users.
+
+        Parameters
+        ----------
+        reco : pd.DataFrame
+            Recommendations table with columns `Columns.User`, `Columns.Item`, `Columns.Rank`.
+
+        Returns
+        -------
+        pd.Series
+            Values of metric (index - user id, values - metric value for every user).
+        """
+        raise NotImplementedError()
+
+
+class RecoEmpty(_RecoDQMetric):
```

Comment:
I wanted to make DQ metrics separated from others even in naming. So I tried to go away from `Completeness` and other names like this because they sound like `Coverage`, 'Accuracy`, etc. Like they are native recommender systems metrics.
But I get your point. We can probably go to:
-`UsersCovered` // **`CoveredUsers`**
-`RecoFull` // `Fulfilled` // `Completed` // `Filled` // `FilledIn` // **`Sufficient`** // `Delivered`
- RecoUnique` // **`Unrepeated`**

Reply:
CoveredUsers, SufficientReco, UnrepeatedReco. Final decision

Reply:
fixed